syntax = "proto3";
import "google/protobuf/any.proto";

enum SerializableDataType{
  UNDEFINED = 0;
  INT8 = 1;
  INT16 = 2;
  INT32 = 3;
  INT64 = 4;
  UINT8 = 5;
  UINT16 = 6;
  UINT32 = 7;
  UINT64 = 8;
}

message SerializableField{
  string name = 1;
  SerializableDataType type = 2;
}

message SerializableExpression {
  repeated SerializableExpression children = 1;
}

message SerializableOperator{

  message SourceDetails {
    message SerializableBinarySourceDescriptor {
      string filePath = 1;
    }

    message SerializableCsvSourceDescriptor {
      string filePath = 1;
      string delimiter = 2;
      uint64 numBuffersToProcess = 3;
      uint64 frequency = 4;
    }

    message SerializableDefaultSourceDescriptor {
      uint64 numBuffersToProcess = 3;
      uint64 frequency = 4;
    }

    message SerializableKafkaSourceDescriptor {
      string brokers = 1;
      string topic = 2;
      string groupId = 3;
      bool autoCommit = 4;
      uint64 kafkaConnectTimeout = 5;
    }

    message SerializableLogicalStreamSourceDescriptor {
      string streamName = 1;
    }

    message SerializableSenseSourceDescriptor {
      string udfs = 1;
    }

    message SerializableZMQSourceDescriptor {
      string host = 1;
      uint32 port = 2;
    }
    google.protobuf.Any sourceDescriptor = 1;
  }

  message SinkDetails {
    SerializableExpression expression = 1;
    SerializableField field = 2;
  }

  message MapDetails {
    SerializableExpression expression = 1;
    SerializableField field = 2;
  }

  message FilterDetails {
    SerializableExpression predicate = 1;
  }

  google.protobuf.Any details = 1;
  repeated SerializableOperator children = 2;
}